### PR TITLE
Explicit Caching

### PR DESCRIPTION
**Pull request by mayureshagashe2105**
Change-Id: I694545243efda467d6fd599beded0dc6679b727d

===DO NOT SUBMIT===

*Inital prototype for explicit caching

*Add basic **CURD** support for caching

*Remove `INPUT_ONLY` marked fields from `CachedContent` dataclass

*Rename files `cached_content*` -> `caching*`

*Update `create` method for explicit instantination of 'CachedContent'

*Add a factory method to instatinate model with `CachedContent` as its context

*blacken

*Add tests

*Rename `get_cached_content` -> `get`

*Stroke out functional approach for `CachedContent` **CURD** ops

*Validate `name` checks for `CachedContent` creation

*Update repr for `GenerativeModel` to include information about the model's context

*Update docstrings

*Improve tests

### CODE REVIEW

**Comment by MarkDaoust** [comment]
Passing, with the preview build.

**Comment by mayureshagashe2105** [comment]
Cool, Thnx!

File: google/generativeai/caching.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+    name: str
+    model: str
+    create_time: datetime.datetime
+    update_time: datetime.datetime
+    expire_time: datetime.datetime
+
+    # NOTE: Automatic CachedContent deletion using contextmanager is not P0(P1+).
+    # Adding basic support for now.
+    def __enter__(self):
```
Comment (MarkDaoust) [inline google/generativeai/caching.py:44]: We're not shipping the context manager for the initial launch. Please remove enter/exit.

File: google/generativeai/client.py [hunk 2]
```diff
@@ -315,6 +315,10 @@ def configure(
 _client_manager.configure()
 
 
+def get_default_cache_client() -> glm.CacheServiceClient:
```
Comment (MarkDaoust) [inline google/generativeai/client.py:318]: Nice.

File: google/generativeai/generative_models.py [hunk 3]
```diff
@@ -165,6 +185,55 @@ def _get_tools_lib(
...
+        Args:
+            cached_content: context for the model.
+
+        Returns:
+            `GenerativeModel` object with `cached_content` as its context.
+        """
+        if isinstance(cached_content, str):
+            cached_content = caching.CachedContent.get(name=cached_content)
+
+        # call __new__ with the cached_content to set the model's context. This is done to avoid
```
Comment (MarkDaoust) [inline google/generativeai/generative_models.py:224]: This thing with __new__ is a little weird.

Why not add it as a public field, default to `None`? A person may want to check which cache they're using.

File: google/generativeai/caching.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+from google.generativeai.types import content_types
+from google.generativeai.utils import flatten_update_paths
+from google.generativeai.client import get_default_cache_client
+
+from google.protobuf import field_mask_pb2
+import google.ai.generativelanguage as glm
+
+
+@dataclasses.dataclass
+class CachedContent:
```
Comment (MarkDaoust) [inline google/generativeai/caching.py:33]: I can't read the "everyting on one line" printout.

please use: https://github.com/google-gemini/generative-ai-python/blob/7b9758f54180b525d355393302100b0dea082a12/google/generativeai/string_utils.py#L52

File: google/generativeai/caching.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+            contents=contents,
+            tools=tools_lib,
+            tool_config=tool_config,
+            ttl=ttl,
+        )
+
+        return protos.CreateCachedContentRequest(cached_content=cached_content)
+
+    @classmethod
+    def create(
```
Comment (MarkDaoust) [inline google/generativeai/caching.py:130]: Async?

**Review by MarkDaoust — APPROVED** [review]

**Comment by MarkDaoust** [comment]
This is passing tests, and works in testing. Let's merge, and we can fix these comments in follow-on PRs.

File: google/generativeai/caching.py (hunk 1 above)
Comment (mayureshagashe2105) [reply google/generativeai/caching.py:44]: Done!

File: google/generativeai/types/caching_types.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+    # NOTE: Automatic CachedContent deletion using contextmanager is not P0(P1+).
+    # Adding basic support for now.
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        self.delete()
+
+    def _to_dict(self) -> glm.CachedContent:
```
[the remark this answers is not in the document]
Comment (mayureshagashe2105) [reply google/generativeai/types/caching_types.py]: `get_update_fields` ?

File: google/generativeai/generative_models.py (hunk 3 above)
Comment (mayureshagashe2105) [reply google/generativeai/generative_models.py:224]: Design doc talked about having exclusive field. But I agree with you, I'll remove the __new__ implementation and add a public property instead.

File: google/generativeai/caching.py (hunk 4 above)
Comment (mayureshagashe2105) [reply google/generativeai/caching.py:33]: Done!

File: google/generativeai/caching.py (hunk 5 above)
Comment (mayureshagashe2105) [reply google/generativeai/caching.py:130]: Didn't implement it, due to the following reason:

1. Does it make sense to have async for cacheservice?
2. Anyway, I'll submit it in a separate PR, just copy paste and async/await :)

**Comment by mayureshagashe2105** [comment]
Thanks for the review 🙌🏻

**Pull request merged**